### ISSUE TITLE
Add support for multi-row INSERT statements

### ISSUE THREAD

**Issue by samay-sharma** [citusdata/citus#455]
Currently, statements like the below error out in Citus:

INSERT INTO counters VALUES ('a','02-01-01',1),('b','01-01-01',2);
ERROR:  cannot perform distributed planning for the given modification
DETAIL:  Multi-row INSERTs to distributed tables are not supported.

Issuing them like below works:

INSERT INTO counters VALUES ('a', '02-01-01',1);
INSERT INTO counters VALUES('b',01-01-01',2);

However, it seems that certain ORMs use the syntax specified above.

# Task List

Broken down by planner, executor, and testing…

## Planning Changes (two days)

 - [x] Remove check preventing multi-row `INSERT`s (0 days)
 - [x] Add logic to group rows by shard (1 day)
 - [x] Change `TargetShardIntervalForModify` to return list of shard/query structs (0.5 days)
 - [x] Change `RouterModifyTask` to expect that as input and produce multiple tasks (0.5 days)

## Executor Changes (one day)

 - [x] Call above logic from `CitusModifyBeginScan` when deferred pruning needed (0 days)
 - [x] Use changes from #1455 to call multi-shard modification code in multi-row case (1 day)

## Testing Changes (1.5 days)

 - [x] Add use of multi-row `INSERT` to reference table, modifying xacts, and insert select test files (0.25 days)
 - [ ] Add isolation tester tests for multi-row feature (1 day)
 - [x] Test interaction with deadlock detection branch (0.25 days)

**Comment by samay-sharma** [comment]
I ran into this in two recent customer conversations.

**Comment by metdos** [comment]
This issue also reported in the citus-users group.

**Comment by btelles** [comment]
+1 This feature would help when implementing Extract, Transform, Load jobs of millions of records where "INSERT INTO counters" adds a bit of overhead. For large workloads and low-column tables like the above, this would save about 25% in bytes transferred (probably more like 10% for larger tables). Calculated this assuming 94 bytes vs 66 bytes for 2 processing two records.
Seems like reasonable gains that are worth implementing.

**Comment by lfittl** [comment]
@btelles I'm curious whether `COPY` (released in 5.1 this week) would work for your use case?

**Comment by julienmarie** [comment]
+1

**Comment by MarioSanzRodrigo** [comment]
Any news about this?

**Comment by ozgune** [comment]
@btelles -- thanks for evaluating Citus and your feedback.

We're looking at this issue as a candidate for our next release cycle. Could you elaborate a bit more on your use-case?

In particular, are you looking to ingest large volumes of raw data into Citus? Or, are you looking to transform and aggregate data in distributed tables?

**Comment by ozgune** [comment]
@julienmarie and @MarioSanzRodrigo -- thanks for posting to this issue.

We're looking at this feature as a candidate for our next release cycle. Could you elaborate a bit more on your use-case for this feature?

**Comment by dimaxgl** [comment]
+1

**Comment by marcocitus** [comment]
For the next release, we are aiming to support multi-row INSERTs across multiple shards. Multi-row inserts can be used to insert a large number of small batches, or a small number of big batches. Ideally, the implementation performs reasonably in both cases.

Design notes:

On the planning side we typically generate 1 task per job in the router planner. We now need to generate a list of tasks for a single INSERT statement. The first step is therefore to refactor functions such as `TargetShardIntervalForModify` and `RouterModifyTask` to generate lists. It might be useful to generalise to `List *QueryTaskList(Query *jobQuery)`. This will also help us resolve cases which prune to 0 shards (#1230), since we can return an empty task list. 

Multi-row INSERTs are special in the sense that we will need to send a different command to every shard. We need to go through each of the rows and find the values in the partition column and the corresponding shard. All rows that map to the same shard should be combined in a single multi-row INSERT query. Once all the rows that go to a particular shard are known, the insert statement can be generated and deparsed.

It might not be possible prune at planning time if the partition column contains function calls or parameters. We can use the the `deferredPruning` flag in the `Job` for that and re-generate the task list by calling `QueryTaskList` from the executor if the flag is set. This should work in the same way.

The tricky part of multi-row INSERTs is execution. We currently don't have an execution code path that can efficiently execute multi-row INSERTs. The router executor can only execute one command on one shard, while the multi-shard router executor opens a connection for each placement and only allows one command in parallel. While the latter is in principle capable of execution a multi-row INSERT and can be used for initial testing, performance would be dreadful.

Adding yet another new executor code path that appropriately handles parallelisation, locking, replication, etc. adds a lot of technical debt. We should aim to create an executor that consolidates one or more of the existing executors with the logic needed to handle multi-row INSERTs.

Additionally, we intend to send `SET lock_timeout TO ...` prior to sending a multi-shard INSERT. Concurrent multi-shard INSERTs might form a distributed deadlock. The lock timeout doesn't prevent deadlock, but at least prevents it from lasting forever. Multi-shard INSERT could also deadlock against COPY commands, `master_modify_multiple_shards`, or `INSERT..SELECT`. We could prevent the latter 2 from running concurrently through a lock on the coordinator.

It would be best to error out if `ON CONFLICT` is set on a multi-shard INSERT, since that makes deadlock considerably more likely.

**Comment by anarazel** [comment]
> Additionally, we intend to send SET lock_timeout TO ... prior to sending a multi-shard INSERT. Concurrent multi-shard INSERTs might form a distributed deadlock. The lock timeout doesn't prevent deadlock, but at least prevents it from lasting forever. Multi-shard INSERT could also deadlock against COPY commands, master_modify_multiple_shards, or INSERT..SELECT. We could prevent the latter 2 from running concurrently through a lock on the coordinator.

This should be a configurable timeout.

> It would be best to error out if ON CONFLICT is set on a multi-shard INSERT, since that makes deadlock considerably more likely.

Indeed.  Alternatively we can have a second, default *off*, setting that allows it with a different timeout (presumably lower).

**Comment by ozgune** [comment]
@btelles, @julienmarie, @MarioSanzRodrigo, and @dimaxgl -- We're moving forward with implementing multi-row INSERTs and wanted to better understand your requirements. If you could comment on the following questions, it would notably help us.

1. Why do you need multi-row ``INSERT``s (compared to some of the workarounds)?
2. How many values do you ``INSERT`` per statement? How many concurrent multi-row ``INSERT``s do you run?
3. Any examples from your write pipeline would also help us a lot.

**Comment by habibc** [comment]
Plus one for this feature. We need this because in our write pathway we process rows in batches of size 32, so it make sense to insert all of them at once. This is more so as we don't have to handle the cases where one batch partially succeeded inserting its data and instead we can fail the whole batch to retry.

**Comment by maxkremer** [comment]
+1
